jetty-ee9/jetty-ee9-nested/src/main/java/org/eclipse/jetty/ee9/nested/HttpChannel.java

```diff
@@ -172,7 +172,10 @@ public boolean needContent()
      */
     public HttpInput.Content produceContent()
     {
-        Content.Chunk chunk = getCoreRequest().read();
+        ContextHandler.CoreContextRequest coreContextRequest = getCoreRequest();
+        if (coreContextRequest == null)
+            return null;
```

Comment:
@lorban is this reasonable?

Reply:
`coreContextRequest` can only be null after `recycle()`, so returning null in that case is wrong as that implies a call to `demand()` should follow.

I think returning an `ErrorContent` would be the right thing to do when `coreContextRequest` is null.

Reply:
I don't see the difference between returning null here, and returning null at line 179 when there is no chunk?

Reply:
`produceContent()` returning null has a special meaning: that means there's no content available right away, but more _will_ follow.

When `coreContextRequest` is null, this means the channel has been recycled, so no more content will follow. In this case, null is not a reasonable return value.

By the way, this `coreContextRequest` null check should also be done in `needContent()` as both `needContent()` and `produceContent()` fulfill the same contract about the availability of content.

Reply:
@lorban it's not possible to return `ErrorContent` for `needContent()`: you can return `false` or throw an `Exception`, which would you prefer?

Reply:
@janbartel as part of my commit, I changed `needContent()` to make it return `true` when `coreContextRequest` is null.

Reply:
@lorban are some of these tests racy? When I run `BlockingTest` in `ee9` and `ee88` locally on my machine, they all pass. When they run in CI, `ee88` all fail with an `IllegalStateException` due to the channel being recycled.  See https://jenkins.webtide.net/blue/organizations/jenkins/jetty.project/detail/PR-12529/13/tests

Reply:
@janbartel Most of those tests were flaky, so I fixed them. But `testBlockingReadAndBlockingWriteGzipped()` was failing on a genuine bug in `HttpChannelState.ChannelResponse` which I fixed too.

But I would appreciate more pairs of eyes (@gregw? @sbordet?) to look at my fix as it's touching delicate code.